CONTRIBUTING.md

```diff
@@ -456,6 +456,172 @@ checks if contributed code does not have the appropriate license headers.
 > be automatically configured to add the correct license header to new source
 > files based on the source location.
 
+### Logging
+
+The Elasticsearch server logs are vitally useful for diagnosing problems in a
+running cluster. You should make sure that your contribution uses logging
+appropriately: log enough detail to inform users about key events and help them
+understand what happened when things go wrong, without logging so much detail
+that the logs fill up with noise and the useful signal is lost.
+
+Elasticsearch uses Log4J for logging. In most cases you should log via a
+`Logger` named after the class that is writing the log messages, which you can
+do by declaring a static field of the class as follows:
+
+    private static final Logger logger = LogManager.getLogger(MyClassName.class);
+
+There are a number of other acceptable ways to obtain a `Logger`. For example,
+if there will be many instances of a long-lived component whose logs should be
+distinguished then each instance can use a prefixed logger obtained from one of
+the methods in `org.elasticsearch.common.logging.Loggers`.
+
+If the log message includes values from your code then you must use use
+placeholders rather than constructing the string yourself. Consider wrapping
+the values in `[...]` to help distinguish them from the static part of the
+message:
+
+    logger.debug("operation failed [{}] times in [{}]ms", failureCount, elapsedMillis);
+
+You can also pass in an exception to log it including its stack trace (and
+causes and so on...):
+
+    logger.debug("operation failed", exception);
+
+If you wish to use placeholders and an exception at the same time, construct a
+`ParameterizedMessage`:
+
+    logger.debug(new ParameterizedMessage("failed at offset [{}]", offset), exception);
+
+You can also use a `Supplier<ParameterizedMessage>` to avoid constructing
+expensive messages that will usually be discarded:
+
+    logger.trace(() -> new ParameterizedMessage("Tolstoy says [{}]", warAndPeace.toString()));
+
+Logging is an important behaviour of the system and all but the most trivial
+logging should have corresponding tests. These tests typically use a
+`org.elasticsearch.test.MockLogAppender` to capture the logs and assert that
+they are as expected.
```

Comment:
This might be true in the abstract, but I've only ever seen/written logging tests for particularly important log messages, or ones that are complex to trigger and I want to make sure are working right. I'm mildly concerned the phrasing here overemphasizes the degree we expect logging tests for most instances of logging.

Reply:
++ I write quite a lot of such tests but I'm willing to believe it's not that common. I'll weaken the language here.

Reply:
I can very easily believe the logging you write falls into the "particularly important or complex to trigger" camp much more often than the logging I write :)